### PR TITLE
when docker image is not present, download it and create in glance

### PR DESCRIPTION
**Pull request by jsuchome**

### CODE REVIEW

**Comment by aplanas** [comment]
+1 Tested in c16

File: chef/cookbooks/tempest/recipes/config.rb [hunk 1]
```diff
@@ -357,7 +358,21 @@
...
+  echo "$DOCKER_IMAGE_ID" > #{docker_image_id_file}
+else
+  for node in #{docker_nodes}; do
+      ssh $node docker pull cirros
+  done
+  DOCKER_IMAGE_ID=$(glance #{insecure} image-create --name #{image_name} \
+      --container-format docker \
+      --disk-format raw --property hypervisor_type=docker \
+      --is-public True  \
+      --copy-from http://clouddata.cloud.suse.de/images/docker/cirros.tar \
```
Comment (vuntz) [inline chef/cookbooks/tempest/recipes/config.rb]: We can't put this URL in the barclamp; this needs to work for external people also.

We could possibly have the URL in an attribute, and then set the attribute in mkcloud, though.

**Comment by vuntz** [comment]
If that works, then fine with me. Note that we do ship the cirros image as part on the admin server, to avoid requiring internet access; the "docker pull" kind of kills this :/

I wonder if there's a way to avoid the "docker pull", though. That will become expensive at some point.

**Comment by vuntz** [comment]
> I wonder if there's a way to avoid the "docker pull", though.

`docker load` is what we want!

**Comment by vuntz** [comment]
So we could also ship the docker version of cirros on the admin server, and do the magic?

File: chef/cookbooks/tempest/recipes/config.rb [hunk 2]
```diff
@@ -357,7 +359,23 @@
     --is-public True \
     --page-size 1 \
     2> /dev/null | tail -n 2 | head -n 1 | awk '{ print $2 }')
-[ -n "$DOCKER_IMAGE_ID" ] && echo "$DOCKER_IMAGE_ID" > #{docker_image_id_file}
+if [ -n "$DOCKER_IMAGE_ID" ] ; then
+  echo "$DOCKER_IMAGE_ID" > #{docker_image_id_file}
+else
+  for node in #{docker_nodes.join(" ")}; do
+      ssh $node docker pull cirros
```
Comment (vuntz) [inline chef/cookbooks/tempest/recipes/config.rb]: Hrm, do we have ssh keys for this?

**Comment by jsuchome** [comment]
> Note that we do ship the cirros image as part on the admin server

That's new to me. Where is that?
But anyway, this will make the barclamp dependent on our admin server again, which I wanted to avoid.

**Comment by jsuchome** [comment]
BTW, speaking of internet access, we're downloading from IMAGE_URL anyway

File: chef/cookbooks/tempest/recipes/config.rb (hunk 2 above)
Comment (jsuchome) [inline chef/cookbooks/tempest/recipes/config.rb]: Oh, that we probably don't have. But cirros image needs to be present on docker nodes.

So maybe again, it is wrong to do it in tempest barclamp and we should rely on mkcloud/other deployment solutions

**Comment by vuntz** [comment]
> That's new to me. Where is that?

See https://github.com/crowbar/barclamp-tempest/blob/master/chef/data_bags/crowbar/bc-template-tempest.json#L7 -- but that's the non-docker cirros. We could add a cirros version, though.

**Comment by vuntz** [comment]
> BTW, speaking of internet access, we're downloading from IMAGE_URL anyway

And by default, IMAGE_URL points to the admin server.

**Comment by jsuchome** [comment]
OK, we can place cirros.tar at admin server, and we can create glance image from that, but that ``docker pull/load`` is still needed at docker compute nodes. How could we access the nodes from the tempest recipe? And should we?

**Comment by jsuchome** [comment]
We may go back to ignoring the situation when docker based glance image is not found. We're doing it for other images in this recipe anyway.

**Comment by vuntz** [comment]
> OK, we can place cirros.tar at admin server, and we can create glance image from that, but that docker pull/load is still needed at docker compute nodes. How could we access the nodes from the tempest recipe? And should we?

One horrible hack would be to change the nova cookbook to check if there's a tempest proposal applied, and if yes, do the `docker pull/load` from the nova recipes...

File: chef/cookbooks/tempest/recipes/config.rb (hunk 2 above)
Comment (dirkmueller) [inline chef/cookbooks/tempest/recipes/config.rb]: I'm not very familiar with how docker works, but wouldn't it be enough to download it from the admin node and upload it to the glance node? why does it need to be on the "docker nodes" ?

Comment (vuntz) [inline chef/cookbooks/tempest/recipes/config.rb]: Nope, that's not enough; there's an explicit limitation that image must be also in the local docker repo of each node right now (which is annoying, indeed).

**Comment by aplanas** [comment]
IMHO this https://github.com/crowbar/barclamp-tempest/pull/162 supersede this PR, but using the docker image from the package.

I will close this.

**Pull request closed without merging**